sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -312,45 +270,460 @@ class reducer<T, BinaryOperation,
+
+  // SYCL 2020 revision 4 says this should be const, but this is a bug
+  // see https://github.com/KhronosGroup/SYCL-Docs/pull/252
+  reducer<T, BinaryOperation, Dims - 1, Extent, Algorithm, true>
+  operator[](size_t Index) {
+    return {MValue[Index], BinaryOperation()};
+  }
+
+  template <typename _T = T, class _BinaryOperation = BinaryOperation>
+  static enable_if_t<has_known_identity_impl<_BinaryOperation, _T>::value, _T>
+  getIdentity() {
+    return known_identity_impl<_BinaryOperation, _T>::value;
+  }
+
+  T &getElement(size_t E) { return MValue[E]; }
+  const T &getElement(size_t E) const { return MValue[E]; }
+
+private:
+  marray<T, Extent> MValue;
+};
+
+/// Base non-template class which is a base class for all reduction
+/// implementation classes. It is needed to detect the reduction classes.
+class reduction_impl_base {};
+
+/// Templated class for common functionality of all reduction implementation
+/// classes.
+template <typename T, class BinaryOperation> class reduction_impl_common {
+protected:
+  reduction_impl_common(const T &Identity, BinaryOperation BinaryOp,
+                        bool Init = false)
+      : MIdentity(Identity), MBinaryOp(BinaryOp), InitializeToIdentity(Init) {}
+
+public:
+  /// Returns the statically known identity value.
+  template <typename _T = T, class _BinaryOperation = BinaryOperation>
+  enable_if_t<IsKnownIdentityOp<_T, _BinaryOperation>::value,
+              _T> constexpr getIdentity() {
+    return known_identity_impl<_BinaryOperation, _T>::value;
+  }
+
+  /// Returns the identity value given by user.
+  template <typename _T = T, class _BinaryOperation = BinaryOperation>
+  enable_if_t<!IsKnownIdentityOp<_T, _BinaryOperation>::value, _T>
+  getIdentity() {
+    return MIdentity;
+  }
+
+  /// Returns the binary operation associated with the reduction.
+  BinaryOperation getBinaryOperation() const { return MBinaryOp; }
+  bool initializeToIdentity() const { return InitializeToIdentity; }
+
+protected:
+  /// Identity of the BinaryOperation.
+  /// The result of BinaryOperation(X, MIdentity) is equal to X for any X.
+  const T MIdentity;
+
+  BinaryOperation MBinaryOp;
+  bool InitializeToIdentity;
+};
+
+/// Types representing specific reduction algorithms
+/// Enables reduction_impl_algo to take additional algorithm-specific templates
+template <bool IsUSM, access::placeholder IsPlaceholder, int AccessorDims>
+class default_reduction_algorithm {};
+
+/// Templated class for implementations of specific reduction algorithms
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
+          class Algorithm>
+class reduction_impl_algo;
+
+/// Original reduction algorithm is the default. It supports both USM and
+/// accessors via a single class
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
+          bool IsUSM, access::placeholder IsPlaceholder, int AccessorDims>
```

Comment:
The parameter 'int AccessorDims' seems redundant as it is always equal to 1 (grep the reduction.hpp files above. 3rd argument is always 1. Grep for "access::placeholder::*' to find those places (or find my comment "Mark1" above as example of such places where AccessorDims is 1). 
Is it needed, or is it reserved for some future usage, or is it added only to not use the magic const 1?

Reply:
Ah, right. legacy reductions allowed 0-dim and 1-dim accessors.